homeassistant/components/matrix/__init__.py

```diff
@@ -291,116 +296,141 @@ def _get_auth_tokens(self) -> JsonObjectType:
-
-        return client
-
-    def _login_by_token(self):
-        """Login using authentication token and return the client."""
-        return MatrixClient(
-            base_url=self._homeserver,
-            token=self._auth_tokens[self._mx_id],
-            user_id=self._mx_id,
-            valid_cert_check=self._verify_tls,
-        )
 
-    def _login_by_password(self):
-        """Login using password authentication and return the client."""
-        _client = MatrixClient(
-            base_url=self._homeserver, valid_cert_check=self._verify_tls
-        )
+        if not self._client.logged_in:
+            raise ConfigEntryAuthFailed(
+                "Login failed, both token and username/password are invalid"
+            )
 
-        _client.login_with_password(self._mx_id, self._password)
+        await self._store_auth_token(self._client.access_token)
 
-        self._store_auth_token(_client.token)
+    async def _send_image(self, image_path: str, target_rooms: list[RoomID]) -> None:
+        """Upload an image, then send it to all target_rooms."""
+        if not self.hass.config.is_allowed_path(image_path):
+            _LOGGER.error("Path not allowed: %s", image_path)
+            return
 
-        return _client
+        # Get required image metadata.
+        image = Image.open(image_path)
+        (width, height) = image.size
+        mime_type = mimetypes.guess_type(image_path)[0]
+        file_stat = await aiofiles.os.stat(image_path)
+
+        _LOGGER.debug("Uploading file from path, %s", image_path)
+        async with aiofiles.open(image_path, "r+b") as image_file:
+            response, _ = await self._client.upload(
+                image_file,
+                content_type=mime_type,
+                filename=os.path.basename(image_path),
+                filesize=file_stat.st_size,
+            )
+        if isinstance(response, UploadError):
+            _LOGGER.error("Unable to upload image to the homeserver: %s", response)
+            return
 
-    def _send_image(self, img, target_rooms):
-        _LOGGER.debug("Uploading file from path, %s", img)
+        assert isinstance(response, UploadResponse)
+        _LOGGER.debug("Successfully uploaded image to the homeserver")
+
+        content = {
+            "body": os.path.basename(image_path),
+            "info": {
+                "size": file_stat.st_size,
+                "mimetype": mime_type,
+                "w": width,
+                "h": height,
+            },
+            "msgtype": "m.image",
+            "url": response.content_uri,
+        }
 
-        if not self.hass.config.is_allowed_path(img):
-            _LOGGER.error("Path not allowed: %s", img)
-            return
-        with open(img, "rb") as upfile:
-            imgfile = upfile.read()
-        content_type = mimetypes.guess_type(img)[0]
-        mxc = self._client.upload(imgfile, content_type)
+        for room in target_rooms:
+            await self._client.room_send(
```

Comment:
Below in the `_send_message` function, you handle `ErrorResponse`. Why not here too?

Reply:
Great point:

I've shuffled the handling around so that `_send_message` and `_send_image` both call `_handle_multi_room_send`, which wraps the `_handle_room_send` coroutines into concurrent tasks.

`_handle_room_send` itself now contains all of the `ErrorResponse` handling, so it's entirely shared between text/html/image sending now.